src/dragonball/src/tracer.rs

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) 2023 Alibaba Cloud
+// Copyright (c) 2023 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+use std::cmp::min;
+use std::fmt::{Debug, Formatter};
+use std::sync::{Arc, Mutex};
+
+use anyhow::Result;
+use log::{info, warn};
+use opentelemetry::global;
+use tokio::runtime::Runtime;
+use tracing::subscriber::DefaultGuard;
+use tracing::{span::EnteredSpan, subscriber::NoSubscriber, trace_span, Subscriber};
+use tracing_opentelemetry::OpenTelemetryLayer;
+use tracing_subscriber::prelude::*;
+use tracing_subscriber::Registry;
+
+const DEFAULT_JAEGER_URL: &str = "http://localhost:14268/api/traces";
+
+/// dragonball tracer error.
+#[derive(Debug, thiserror::Error)]
+pub enum DragonballTraceError {
+    /// verify jaeger config failed.
+    #[error("config verify failed: {0}")]
+    ConfigVerifyFailed(anyhow::Error),
+}
+
+/// Configuration information for tracing feature.
+#[derive(Clone)]
+pub struct TraceConfigInfo {
+    /// subscriber from runtime.
+    pub subscriber: Option<Arc<dyn Subscriber + Send + Sync>>,
+    /// Tracer name.
+    pub sid: String,
+    /// If enabled, the hpyervisor will create opentracing.io traces and spans.
+    /// See https://www.jaegertracing.io/docs/getting-started.
+    /// The full url to the Jaeger HTTP Thrift collector.
+    pub jaeger_endpoint: String,
+
+    /// The username to be used if basic auth is required for Jaeger.
+    pub jaeger_user: String,
+
+    /// The password to be used if basic auth is required for Jaeger.
+    pub jaeger_password: String,
+}
+
+impl Debug for TraceConfigInfo {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TraceConfigInfo")
+            .field("sid", &self.sid)
+            .field("jaeger_endpoint", &self.jaeger_endpoint)
+            .field("jaeger_user", &self.jaeger_user)
+            .field("jaeger_password", &self.jaeger_password)
+            .finish()
+    }
+}
+
+impl PartialEq for TraceConfigInfo {
+    fn eq(&self, other: &Self) -> bool {
+        if self.sid == other.sid
+            && self.jaeger_endpoint == other.jaeger_endpoint
+            && self.jaeger_user == other.jaeger_user
+            && self.jaeger_password == other.jaeger_password
+        {
+            return true;
+        }
+        false
+    }
+}
+
+impl Eq for TraceConfigInfo {}
```

Comment:
use derive

Reply:
It will encounter a compile error:  the trait `std::cmp::Eq` is not implemented for `(dyn tracing::Subscriber + Send + Sync + 'static)`